url3f8d6440dd_huiwq1990_CognativeModel_tgz-pJ8-clarion_samples_HelloWorldJ8/src/clarion/system/CLARION.java

```diff
@@ -0,0 +1,783 @@
+		
+		if(CurrentInput == null)
+			throw new MissingSensoryInformationException ("Sensory Information has not been provided to " +
+					"this CLARION Agent. Please provide the agent with sensory information by calling the " +
+					"perceive method before calling the act method.");
+		
+		if(mcs != null)
+		{
+			Collection <AbstractMetaCognitiveModule> m = mcs.getModules(InterfaceMCSRunsBeforeACS.class);
+			if(m != null)
+			{
+				for(AbstractMetaCognitiveModule a : m)
+				{
+					if(a.checkEligibility())
+						a.performMetaCognition(CurrentInput, ts);
+				}
+			}
+		}
+		
+		AbstractAction ChosenAction = acs.chooseAction(CurrentInput, ts);
+		if(em != null)
+		{
+			em.add(CurrentInput, ts, InfoStored.CURRENT_STATE);
+			em.add(ChosenAction.clone(), ts, InfoStored.PERFORMED_ACTION);
+		}
+		if(ms != null)
+		{
+			Collection <Drive> dc = ms.getAllDrives();
+			if(dc != null)
+			{
+				for(Drive d : dc)
+					d.updateDeficit();
+			}
+		}
+		if(mcs != null)
+		{
+			Collection <AbstractMetaCognitiveModule> m = mcs.getModules(InterfaceMCSRunsAfterACS.class);
+			if(m != null)
+			{
+				for(AbstractMetaCognitiveModule a : m)
+				{
+					if(a.checkEligibility())
+						a.performMetaCognition(CurrentInput, ts);
+				}
+			}
+		}
+		if(em != null)
+		{
+			ChosenAction = em.getPerformedAction(ts);
+			ChosenAction.setActivation(PerformedAction.FULL_ACTIVATION_LEVEL);
+			PerformedAction = acs.getPossibleActions().get(ChosenAction.getID());
+		}
+		else
+			PerformedAction = ChosenAction;
+		PerformedAction.addTimeStamp(ts);
+		for(Dimension d : CurrentInput.values())
+		{
+			if(d.getNumActivatedVals() > 0)
+			{
+				Dimension ind = InputSpace.get(d.getID());
+				ind.addTimeStamp(ts);
+				for(Value v: d.values())
+				{
+					if(v.isActivated())
+					{
+						ind.get(v.getID()).addTimeStamp(ts);
+					}
+				}
+			}
+		}
+		return PerformedAction;
+	}
+	
+	/**
+	 * Generates the pre-input to use inside of CLARION for meta-cognition performed in the
+	 * early stages of perception.
+	 * @param NewSensoryInformation The new sensory information being perceived.
+	 * @return The pre-input for use within CLARION.
+	 */
+	private DimensionValueCollection generatePreInput (DimensionValueCollection NewSensoryInformation)
+	{
+		DimensionValueCollection preInput = NewSensoryInformation.clone();
+		if(ms != null)
+		{
+			preInput.putAll(ms.calculateAllDriveStrengths(NewSensoryInformation).toDimensionValueCollection());
+		}
+		return preInput;
+	}
+	
+	/**
+	 * Generates the current input to use inside of CLARION given the current sensory information,
+	 * the current goal, the episodic memory, the MS, and the items in working memory.
+	 * @param CurrentSensoryInformation The current sensory information being perceived.
+	 * @return The current input for use within CLARION.
+	 */
+	private DimensionValueCollection generateCurrentInput (DimensionValueCollection CurrentSensoryInformation)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 30. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 377-395, lines 465-493 into separate methods.